wfview: init at latest

jake-arkinstall opened the pull request:
## Description of changes

- Added myself as a maintainer
- Added qcustomplot, a Qt widget for plotting 
- Added wfview, an application for using icom amateur radios from the desktop

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jake-arkinstall committed:
maintainers: add jake-arkinstall

jake-arkinstall committed:
qcustomplot: init at 2.1.1

This is a plotting library used by various Qt applications, and is
available in debian. Debian's implementation includes a patch file
that adds cmake files to qcustomplot, allowing it to be built easily.
I have adopted that patch in this commit, and referenced the Debian
source as appropriate.

My own motivation for requiring it in nixpkgs is that I require it
for wfview in a following commit.

jake-arkinstall committed:
wfview: init at master

wfview is a handy application for Icom radio tranceiver users. It serves as
an open-source alternative to Icom's proprietary application. The
install process is designed for ubuntu-like systems so I have added a
nix install approach.

jake-arkinstall committed:
Removed erroneous whitespace in all-packages.nix

jake-arkinstall committed:
Removed erroneous whitespace in all-packages.nix (again)

sikmir commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31853,6 +31853,8 @@ with pkgs;
 
   welle-io = libsForQt5.callPackage ../applications/radio/welle-io { };
 
+  wfview = callPackage ../applications/radio/wfview { };
```
```suggestion
  wfview = libsForQt5.callPackage ../applications/radio/wfview { };
```

sikmir commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34897,6 +34899,8 @@ with pkgs;
 
   qcomicbook = libsForQt5.callPackage ../applications/graphics/qcomicbook { };
 
+  qcustomplot = callPackage ../applications/graphics/qcustomplot { };
```
```suggestion
  qcustomplot = libsForQt5.callPackage ../applications/graphics/qcustomplot { };
```

sikmir commented on pkgs/applications/radio/wfview/default.nix:
```diff
@@ -0,0 +1,66 @@
+  src = fetchFromGitLab {
+    owner = "eliggett";
+    repo = "wfview";
+    rev = "b68874d32878278e6e84668aaa44534255e7e1f5";
```
Why not [v1.64](https://gitlab.com/eliggett/wfview/-/tags/v1.64)?

sikmir commented on pkgs/applications/radio/wfview/default.nix:
```diff
@@ -0,0 +1,66 @@
+{ stdenv, lib, fetchFromGitLab, qcustomplot, qt5, libopus, eigen, portaudio
+, rtaudio, hidapi, udev, pulseaudio }:
+stdenv.mkDerivation {
+  name = "wfview";
```
```suggestion
  pname = "wfview";
```
`version` is missed.

jake-arkinstall committed:
Address suggestions by @sikmir

Set wfview to 1.64 rather than the latest commit as suggested, and used
libsForQt5.callPackage rather than callPackage.

sikmir commented on pkgs/applications/graphics/qcustomplot/default.nix:
```diff
@@ -0,0 +1,34 @@
+  };
+  patches = [ ./add-cmake.patch ];
+  cmakeFlags = [ "-DUSE_QT_VERSION=5" ];
+  nativeBuildInputs = [ gcc cmake pkg-config ];
```
```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
```

sikmir commented on pkgs/applications/graphics/qcustomplot/default.nix:
```diff
@@ -0,0 +1,34 @@
+{ stdenv, lib, fetchurl, qt5, gcc, cmake, pkg-config }:
+stdenv.mkDerivation {
```
```suggestion
stdenv.mkDerivation rec {
```

jake-arkinstall committed:
Address further changes suggested by @sikmir

AndersonTorres commented on pkgs/applications/graphics/qcustomplot/add-cmake.patch:
```diff
@@ -0,0 +1,146 @@
+Borrowed with love from Debian: https://salsa.debian.org/science-team/qcustomplot/-/blob/master/debian/patches/01_add_CMake_config_material.patch
```
Why?

And why not to `fetchpatch` this patch instead of vendoring it?

AndersonTorres commented on pkgs/applications/graphics/qcustomplot/default.nix:
```diff
@@ -0,0 +1,34 @@
+{ stdenv, lib, fetchurl, qt5, gcc, cmake, pkg-config }:
```
Why gcc?

AndersonTorres commented on pkgs/applications/graphics/qcustomplot/default.nix:
```diff
@@ -0,0 +1,34 @@
+    sha256 = "sha256:Xi0i3sd5248B81fL2yXlT7z5ca2u516ujXrSRESHGC8=";
+  };
+  patches = [ ./add-cmake.patch ];
+  cmakeFlags = [ "-DUSE_QT_VERSION=5" ];
```
I have found that Archlinux does not use cmake, but qmake.

https://gitlab.archlinux.org/archlinux/packaging/packages/qcustomplot/-/blob/main/PKGBUILD?ref_type=heads

Remove the cmake patches - let's respect upstream.

AndersonTorres commented on pkgs/applications/graphics/qcustomplot/default.nix:
```diff
@@ -0,0 +1,34 @@
+  nativeBuildInputs = [ cmake pkg-config ];
+  dontWrapQtApps = true;
+  buildInputs = [ qt5.qtbase ];
+  meta = with lib; {
+    homepage = "https://qtcustomplot.com/";
+    description = "A Qt C++ widget for plotting and data visualization";
+    license = licenses.gpl3;
+
+    longDescription = ''
+      QCustomPlot is a Qt C++ widget for plotting and data visualization.
+      It has no further dependencies and is well documented. This plotting
+      library focuses on making good looking, publication quality 2D plots,
+      graphs and charts, as well as offering high performance for realtime
+      visualization applications. QCustomPlot can export to various formats
+      such as vectorized PDF files and rasterized images like PNG, JPG and BMP.
+      QCustomPlot is the solution for displaying of realtime data inside the
+      application as well as producing high quality plots for other media.
+    '';
+
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ jake-arkinstall ];
+  };
+
```
```suggestion

  meta = {
    homepage = "https://qtcustomplot.com/";
    description = "A Qt C++ widget for plotting and data visualization";
    longDescription = ''
      QCustomPlot is a Qt C++ widget for plotting and data visualization.
      It has no further dependencies and is well documented. This plotting
      library focuses on making good looking, publication quality 2D plots,
      graphs and charts, as well as offering high performance for realtime
      visualization applications. QCustomPlot can export to various formats
      such as vectorized PDF files and rasterized images like PNG, JPG and BMP.
      QCustomPlot is the solution for displaying of realtime data inside the
      application as well as producing high quality plots for other media.
    '';
    license = lib.licenses.gpl3Plus;
    maintainers = with lib.maintainers; [ jake-arkinstall ];
    platforms = lib.platforms.linux;
  };

```

AndersonTorres commented on pkgs/applications/radio/wfview/default.nix:
```diff
@@ -0,0 +1,66 @@
+    rev = "v${version}";
+    sha256 = "QdrCji+gktrWfJKf9+yPkmdzfXOMIh4lwG2JA8WVUT8=";
+  };
+  patches = [ ./remove-syscalls.patch ];
```
Why?

jake-arkinstall replied:
This was a fun one.

The upstream build script uses system calls to gather information about the system to bake into the build for informational purposes (calls directly to hostname, whoami, and git to grab the version). The first two violate hermeticity, and the git call assumes the build is being performed from within a git repository rather than an archive tarball.

The upstream build script also invokes /sbin/ldconfig directly to search for qcustomplot to determine how to link it, due to varying capitalisation options. In the Nix build, we already have qcustomplot on the link path and know how to link it.

The patch resolves this by removing those calls and replacing them with placeholders.

But it does highlight that a comment would be handy to explain why it's there. Additionally, I'll see if I can inject the git version in from the derivation rather than using a placeholder.

AndersonTorres commented on pkgs/applications/radio/wfview/default.nix:
```diff
@@ -0,0 +1,66 @@
+    owner = "eliggett";
+    repo = "wfview";
+    rev = "v${version}";
+    sha256 = "QdrCji+gktrWfJKf9+yPkmdzfXOMIh4lwG2JA8WVUT8=";
```
```suggestion
    hash = "QdrCji+gktrWfJKf9+yPkmdzfXOMIh4lwG2JA8WVUT8=";
```
